src/Bicep.RegistryModuleTool/ModuleFiles/MainArmTemplateFile.cs

```diff
@@ -98,34 +118,28 @@ public MainArmTemplateFile WriteToFileSystem(IFileSystem fileSystem)
             return this;
         }
 
-        private static MainArmTemplateParameter ToParameter(JsonProperty parameterProperty)
+        private MainArmTemplateParameter ToParameter(JsonProperty parameterProperty)
         {
-            string name = parameterProperty.Name;
-            string type = parameterProperty.Value.GetProperty("type").ToNonNullString();
-            bool required = !parameterProperty.Value.TryGetProperty("defaultValue", out _);
-            string? description = TryGetDescription(parameterProperty.Value);
+            var parameters = this.armTemplate.Parameters;
+
+            string name = parameters[parameterProperty.Name].Name;
+            string type = GetPrimitiveTypeName(parameters[parameterProperty.Name].TypeReference);
+            bool required = parameters[parameterProperty.Name].IsRequired;
+            string? description = parameters[parameterProperty.Name].Description;
 
             return new(name, type, required, description);
         }
 
-        private static MainArmTemplateOutput ToOutput(JsonProperty outputProperty)
+        private MainArmTemplateOutput ToOutput(JsonProperty outputProperty)
         {
-            string name = outputProperty.Name;
-            string type = outputProperty.Value.GetProperty("type").ToNonNullString();
-            string? description = TryGetDescription(outputProperty.Value);
 
-            return new(name, type, description);
-        }
+var outputs = this.armTemplate.Outputs.ToImmutableDictionaryExcludingNullValues(x => x.Name, StringComparer.Ordinal.);
```

Comment:
@shenglol , this line still feels fishy to me. I could move OutputsDict is a class variable, instead of performing this conversion every time the method is called.
Any other suggestions?

Reply:
You can search the outputs array instead of converting it to a dictionary:

```csharp
var output = armTemplate.Outputs.Where(x => StringComparer.Ordinal.Equals(x.Name, outputProperty.Name)).Single();
return new(output.Name, GetPrimitiveType(output.TypeReference), output.Description);
```

It could be more efficient, but I think it's fine to leave this left than perfect and then change `ISemanticModel.Outputs` to be a dictionary in a future PR.

Reply:
I think if we go with `Where.` and  `.First()` instead of `.Single()` we can gain some efficiency.

Reply:
started throwing errors when i tried this, im just going to leave what we had. (i think its upset in this case if outputProperty is empty)